arraycontext/impl/pytato.py

```diff
@@ -0,0 +1,536 @@
+        import pyopencl.array as cla
+        from arraycontext import (is_array_container,
+                                  rec_keyed_map_array_container)
+
+        input_kwargs_to_loopy = {}
+
+        # {{{ extract loopy arguments execute the program
+
+        for pos, arg in enumerate(args):
+            if isinstance(arg, np.number):
+
+                input_kwargs_to_loopy[self.input_id_to_name_in_program[(pos,)]] = (
+                    cla.to_device(self.actx.queue, np.array(arg)))
+            elif is_array_container(arg):
+                def _extract_lpy_kwargs(keys, ary):
+                    if isinstance(ary, pt.array.DataWrapper):
+                        processed_ary = ary.data
+                    elif isinstance(ary, cla.Array):
+                        processed_ary = ary
+                    elif isinstance(ary, pt.Array):
+                        processed_ary = (self.actx.freeze(ary)
+                                         .with_queue(self.actx.queue))
+                    else:
+                        raise TypeError("Expect pt.Array or CL-array, got "
+                                f"{type(ary)}")
+
+                    input_kwargs_to_loopy[
+                        self.input_id_to_name_in_program[(pos,)
+                                                         + keys]] = processed_ary
+                    return ary
+
+                rec_keyed_map_array_container(_extract_lpy_kwargs, arg)
+            else:
+                raise NotImplementedError(type(arg))
+
+        # {{{ the generated program might not have depended on some of the
+        # inputs => do not pass those to the loopy kernel
+
+        input_kwargs_to_loopy = {arg_name: arg
+                                 for arg_name, arg in input_kwargs_to_loopy.items()
+                                 if arg_name in (self.pytato_program
+                                                 .program.default_entrypoint
+                                                 .arg_dict)}
+
+        # }}}
+
+        evt, out_dict = self.pytato_program(queue=self.actx.queue,
+                                            allocator=self.actx.allocator,
+                                            **input_kwargs_to_loopy)
+
+        evt.wait()
+
+        # }}}
+
+        def to_output_template(keys, _):
+            return self.actx.thaw(out_dict[self.output_id_to_name_in_program[keys]])
+
+        return rec_keyed_map_array_container(to_output_template,
+                                             self.output_template)
+
+
+class PytatoArrayContext(ArrayContext):
+    """
+    A :class:`ArrayContext` that uses :mod:`pytato` data types to represent
+    the DOF arrays targeting OpenCL for offloading operations.
+
+    .. attribute:: queue
+
+        A :class:`pyopencl.CommandQueue`.
+
+    .. attribute:: allocator
+
+        A :mod:`pyopencl` memory allocator. Can also be None (default) or False
+        to use the default allocator.
+    """
+
```

Comment:
We should discuss our strategy for handling of scalars. It seems that the current strategy here is for `PyOpenCLArrayContext` to keep returning plain scalars, where as the pytato array context will return lazy scalars. This is kind of inconsistent, but it might be the best strategy available to us. In the medium term, I would like it if `PyOpenCLArrayContext` started returning device scalars as well, but that's a large-scale compatibility break. #40 for the pyopencl end of this. @kaushikcfd @matthiasdiener basically what I'm looking for here from you is whether you agree with this direction.

Reply:
I agree, I think we should allow both types of scalar i.e. host and device for all operations. But, we should allow a flag which doesn't return host-scalar unless to_numpy-ed.

Reply:
I think allowing both types just complicates things. `to_numpy` isn't too hard to type.